apps/payroll/contracts/Payroll.sol

```diff
@@ -0,0 +1,586 @@
+     * @param accountAddress Employer's address to receive Payroll
+     * @param allowedTokens Array of tokens allowed for payment
+     * @param initialYearlyUSDSalary Employee's salary
+     */
+    function addEmployee(
+        address accountAddress,
+        address[] allowedTokens,
+        uint256 initialYearlyUSDSalary
+    )
+        public
+        auth(OWNER_WRITE_ROLE)
+    {
+        _addEmployeeWithName(
+            accountAddress,
+            allowedTokens,
+            initialYearlyUSDSalary,
+            ""
+        );
+    }
+
+    /**
+     * @dev Add employee to Payroll
+     * @notice Creates employee, adds it to mappings, initializes values
+               and tries to update allowed tokens exchange rates if needed.
+               Updates also global Payroll salary sum.
+     * @param accountAddress Employer's address to receive Payroll
+     * @param allowedTokens Array of tokens allowed for payment
+     * @param initialYearlyUSDSalary Employee's salary
+     * @param name Employee's name
+     */
+    function addEmployeeWithName(
+        address accountAddress,
+        address[] allowedTokens,
+        uint256 initialYearlyUSDSalary,
+        string name
+    )
+        public
+        auth(OWNER_WRITE_ROLE)
+    {
+        _addEmployeeWithName(
+            accountAddress,
+            allowedTokens,
+            initialYearlyUSDSalary,
+            name
+        );
+    }
+
+    /**
+     * @dev Set employee's annual salary
+     * @notice Updates also global Payroll salary sum
+     * @param employeeId Employee's identifier
+     * @param yearlyUSDSalary Employee's new salary
+     */
+    function setEmployeeSalary(uint256 employeeId, uint256 yearlyUSDSalary) public auth(OWNER_WRITE_ROLE) {
+        /* check that employee exists */
+        require(employeeIds[employees[employeeId].accountAddress] != 0);
+
+        yearlyTotalPayroll = yearlyTotalPayroll.sub(employees[employeeId].yearlyUSDSalary);
+        employees[employeeId].yearlyUSDSalary = yearlyUSDSalary;
+        yearlyTotalPayroll = yearlyTotalPayroll.add(yearlyUSDSalary);
+    }
+
+    /**
+     * @dev Remove employee from Payroll
+     * @notice Updates also global Payroll salary sum
+     * @param employeeId Employee's identifier
+     */
+    function removeEmployee(uint256 employeeId) public auth(OWNER_WRITE_ROLE) {
+        /* check that employee exists */
+        require(employeeIds[employees[employeeId].accountAddress] != 0);
+
+        yearlyTotalPayroll = yearlyTotalPayroll.sub(employees[employeeId].yearlyUSDSalary);
+        delete employeeIds[employees[employeeId].accountAddress];
+        delete employees[employeeId];
+        numEmployees--;
+    }
+
+    /**
+     * @dev Payable function to receive ETH
+     * @notice Payable function to receive ETH
+     */
+    function addFunds() public payable {
+        // convert ETH to EtherToken
+        etherToken.wrapAndCall.value(msg.value)(address(finance), "Adding Funds");
```

Comment:
We can wrap `this.balance` better than `msg.value` to ensure that if there was ETH stuck in the contract for some reason, it goes to the vault too!

Reply:
After that we can always `assert(this.balance == 0)`

Reply:
Good idea!